File: Coder Army/SOLID/ISP(interfaceSegregationPrinciple)/Javascript code/ISP_followed.js
class TwoDShape{
  area(){}
}

class ThreeDShape{
  area(){}
  volume(){}
}

class Square extends TwoDShape{
  constructor(side){
    super()
    this.side=side
  }
  area(){
    return this.side*this.side
  }
}

class Rectangle extends TwoDShape{
  constructor(length, breadth){
    super()
    this.length=length
    this.breadth= breadth
  }
  area(){
    return this.length*this.breadth
  }
}

class Cube extends ThreeDShape{
  constructor(side){
    super()
    this.side=side
  }
  area(){
    return 6* this.side*this.side
  }
  volume(){
    return this.side*this.side*this.side
  }
}

const square= new Square(4);
const rectangle= new Rectangle(9, 8)
const cube= new Cube(5);

console.log("Square Area:", square.area());
try {
  console.log("Square Volume:", square.volume());
} catch (e) {
  console.error("Square Volume Error:", e.message);
}

console.log("Rectangle Area:", rectangle.area());
try {
  console.log("Rectangle Volume:", rectangle.volume());
} catch (e) {
  console.error("Rectangle Volume Error:", e.message);
}

console.log("Cube Area:", cube.area());
console.log("Cube Volume:", cube.volume());

File: Coder Army/SOLID/ISP(interfaceSegregationPrinciple)/C++ code/ISP_followed.cpp
#include <iostream>
using namespace std;

class TwoDShape
{
public:
    virtual double area() = 0;
};

class ThreeDShape
{
public:
    virtual double area() = 0;
    virtual double volume() = 0;
};

class Square : public TwoDShape
{
private:
    double side;

public:
    Square(double side) : side(side) {}

    double area() override
    {
        return side * side;
    }
};

class Rectangle : public TwoDShape
{
private:
    double length, width;

public:
    Rectangle(double length, double width) : length(length), width(width) {}

    double area() override
    {
        return length * width;
    }
};

class Cube : public ThreeDShape
{
private:
    double side;

public:
    Cube(double side) : side(side) {}

    double area() override
    {
        return 6 * side * side;
    }

    double volume() override
    {
        return side * side * side;
    }
};

int main()
{
    TwoDShape *square = new Square(5);
    TwoDShape *rectangle = new Rectangle(3, 4);
    ThreeDShape *cube = new Cube(5);

    cout << " area of Square " << square->area() << endl;
    cout << " area of rectangle " << rectangle->area() << endl;
    cout << " area of cube " << cube->area() << endl;
    cout << " volume of cube " << cube->volume() << endl;

    delete square;
    delete rectangle;
    delete cube;
    return 0;
}